src/base/utils/string.cpp

```diff
@@ -130,6 +130,20 @@ bool NaturalCompare::lessThan(const QString &left, const QString &right)
     return false;
 }
 
+bool Utils::String::naturalCompareCaseSensitive(const QString &left, const QString &right)
+{
+    // provide a single `NaturalCompare` instance for easy use
+    static NaturalCompare nCmp(true);  // this is thread-safe in C++11 (stated in spec 6.7.4)
+    return nCmp(left, right);
+}
+
+bool Utils::String::naturalCompareCaseInsensitive(const QString &left, const QString &right)
+{
+    // provide a single `NaturalCompare` instance for easy use
+    static NaturalCompare nCmp(false);  // this is thread-safe in C++11 (stated in spec 6.7.4)
```

Comment:
>But if QCollator is implicitly shared its copying doesn't bring thread-safety here since copies use the same shared data.

This argument violates the idea of "copy-on-write"...
anyway the source is here: https://github.com/qtproject/qtbase/tree/fb6000a74f57bd3c096f6a10142477bf2faf0ff2/src/corelib/tools

look for `QCollator::compare` in qcollator_posix.cpp, qcollator_icu.cpp ...